Applications/ConsoleReferenceClient/UAClient.cs

```diff
@@ -124,19 +138,46 @@ public async Task<bool> ConnectAsync(string serverUrl, bool useSecurity = true)
                 }
                 else
                 {
-                    m_output.WriteLine("Connecting to... {0}", serverUrl);
+                    ITransportWaitingConnection connection = null;
+                    EndpointDescription endpointDescription = null;
+                    if (m_reverseConnectManager != null)
+                    {
+                        m_output.WriteLine("Waiting for reverse connection to.... {0}", serverUrl);
+                        do
+                        {
+                            using (var cts = new CancellationTokenSource(30_000))
+                            {
+                                connection = await m_reverseConnectManager.WaitForConnection(new Uri(serverUrl), null, cts.Token);
+                                if (connection == null)
+                                {
+                                    throw new ServiceResultException(StatusCodes.BadTimeout, "Waiting for a reverse connection timed out.");
+                                }
+                                if (endpointDescription == null)
+                                {
+                                    m_output.WriteLine("Discover reverse connection endpoints....");
+                                    endpointDescription = CoreClientUtils.SelectEndpoint(m_configuration, connection, useSecurity);
+                                    connection = null;
+                                }
+                            }
+                        } while (connection == null);
+                    }
+                    else
+                    {
+                        m_output.WriteLine("Connecting to... {0}", serverUrl);
+                        endpointDescription = CoreClientUtils.SelectEndpoint(m_configuration, serverUrl, useSecurity);
```

Comment:
Would be great to have a verson of SelectEndpoint that takes a reverseConnectManager reference or when null uses the regular mode.  Could be async ideally.